File: ddt/scoring.py
from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import date
from typing import Optional


def clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))


@dataclass(frozen=True)
class DebtBreakdown:
    debt: int
    days_open: int
    overdue_days: int
    severity: float
    growth: float
    overdue_boost: float


def compute_debt(
    *,
    created_date: date,
    due_date: Optional[date],
    impact: int,
    stress: int,
    today: date,
) -> DebtBreakdown:
    """
    Debt score: 0..100

    - growth = 1 - exp(-days_open/21)
    - severity = (impact + stress)/10  in [0.2..1.0]
    - overdue_boost = 1 + min(0.5, overdue_days/30)
    - debt = round(clamp(100 * growth * severity * overdue_boost, 0, 100))
    """
    days_open = max(0, (today - created_date).days)

    if not (1 <= impact <= 5):
        raise ValueError("impact must be between 1 and 5")
    if not (1 <= stress <= 5):
        raise ValueError("stress must be between 1 and 5")

    severity = (impact + stress) / 10.0
    growth = 1.0 - math.exp(-days_open / 21.0)

    overdue_days = 0
    if due_date is not None:
        overdue_days = max(0, (today - due_date).days)

    overdue_boost = 1.0 + min(0.5, overdue_days / 30.0)

    raw_debt = 100.0 * growth * severity * overdue_boost
    debt = int(round(clamp(raw_debt, 0.0, 100.0)))

    return DebtBreakdown(
        debt=debt,
        days_open=days_open,
        overdue_days=overdue_days,
        severity=severity,
        growth=growth,
        overdue_boost=overdue_boost,
    )


def compute_health_score(total_debt: int) -> int:
    """
    Health score: 0..100, higher is better.

    health = round(100 * exp(-total_debt/200))
    """
    if total_debt < 0:
        total_debt = 0
    health = 100.0 * math.exp(-total_debt / 200.0)
    return int(round(clamp(health, 0.0, 100.0)))

File: web/lib/scoring.ts
export type DebtBreakdown = {
  debt: number; // 0-100
  daysOpen: number;
  overdueDays: number;
};

function clamp(n: number, min: number, max: number) {
  return Math.max(min, Math.min(max, n));
}

function daysBetween(a: Date, b: Date) {
  const ms = b.getTime() - a.getTime();
  return Math.floor(ms / (1000 * 60 * 60 * 24));
}

// A smooth compounding-ish curve that doesn’t explode:
// baseAge = 1 - exp(-daysOpen / 21)  (fast early growth then flattens)
function ageFactor(daysOpen: number) {
  return 1 - Math.exp(-daysOpen / 21);
}

// Overdue adds extra pressure but still bounded
function overdueBoost(overdueDays: number) {
  if (overdueDays <= 0) return 0;
  return 1 - Math.exp(-overdueDays / 14);
}

export function computeDebt(params: {
  createdDate: string; // YYYY-MM-DD
  dueDate?: string | null; // YYYY-MM-DD
  impact: number; // 1-5
  stress: number; // 1-5
  today?: Date;
}): DebtBreakdown {
  const today = params.today ?? new Date();

  const created = new Date(params.createdDate + "T00:00:00");
  const daysOpen = Math.max(0, daysBetween(created, today));

  const due = params.dueDate ? new Date(params.dueDate + "T00:00:00") : null;
  const overdueDays = due ? Math.max(0, daysBetween(due, today)) : 0;

  // Multipliers: impact & stress influence but don’t dominate
  const impactMult = 0.75 + params.impact * 0.10; // 0.85..1.25
  const stressMult = 0.75 + params.stress * 0.10; // 0.85..1.25

  const base = ageFactor(daysOpen); // 0..~1
  const overdue = overdueBoost(overdueDays); // 0..~1

  // Combine: base is the core; overdue accelerates if late
  // Weighted sum then scaled to 0..100
  const raw = (0.78 * base + 0.22 * overdue) * impactMult * stressMult;

  const debt = clamp(Math.round(raw * 100), 0, 100);

  return { debt, daysOpen, overdueDays };
}

export function computeHealthScore(totalDebt: number) {
  // Total debt can exceed 100, health should remain 0..100
  // Simple interpretation: 0 debt => 100 health, 400 debt => near 0
  const score = Math.round(100 * Math.exp(-totalDebt / 160));
  return clamp(score, 0, 100);
}
